docs/proposals/component-registration.md

```diff
@@ -0,0 +1,406 @@
+
+Note: Self-healing external components requires knowing how to deploy the external components, which is different for each cloud platform and changes based on component configuration. This is different than self-healing addon components, which is covered by the Kubernetes replication controller.
+
+
+## Design
+
+Unfortunately, ComponentStatuses is almost not worth the effort to fix. Instead it can be deprecated and replaced by a combination of Service and Endpoints changes. This allows the components and their addresses & ports to be specified more flexibly.
+
+Since it's seemingly impossible to fix Endpoints Problem \#1 without breaking reverse compatibility, this proposal punts on \#1 (until v2 allows reverse-incompatibility) and focuses on fixing \#2 for now. #13778 handles increasing the scope of Endpoints to cover both Ready and NotReady addresses, which is better than nothing, but not as flexible as having an array of conditions, like other -Status API resources.
+
+To fix Endpoints Problem \#2, this proposal moves the specification (and status) of external service provider addresses to its own namespaced API resource (provisionally called **ExternalServiceProvider**). Endpoints Controller will then continue to manage the lifecycle (create/update/delete) of Internal Endpoints based on a selection of Pod instances, but will also do the same for External Endpoints based on a selection of ExternalServiceProviders defined in ServiceSpec.Selector. Endpoints then becomes a (conceptually) read-only API for the user. Endpoints can't actually be read-only, because the Endpoints Controller needs access, but write access may eventually be limited to admin/operator-only.
+
+
+### Phased Changes
+
+Since the changes here are complex, they will be implemented in phases.
+
+1. Simplify Endpoints
+  - Add a new API resource (ExternalServiceProvider) that represents a non-pod service provider
+    - Specify a network route (host[:port])
+    - Specify liveness and readiness probe methods
+    - Store the last known status (list of conditions)
+  - Add (or modify) a controller to probe ExternalServiceProviders and update their status
+  - Update how Endpoints are managed
+    - EITHER update the EndpointsController to manage Endpoints based on both Pod and ExternalServiceProvider status
+    - OR modify Endpoints to be a read-only view of service providers (Pod/ExternalServiceProvider) and delete the Endpoints controller
+2. Register component ExternalServiceProviders
+  - Update all components to use consistent namespacing and component labeling
+  - EITHER update "external" components to self-register their ExternalServiceProvider
+  - OR update deployment automation to register deployed components as ExternalServiceProviders
+3. Rewrite ComponentStatuses
+  - Dynamically display a list of components and their readiness (filtered Pod/ExternalServiceProvider or Endpoints query) on `/componentstatuses`
```

Comment:
I think making a new resource & new, api conforming location for it is better -- we can leave the old thing until all clients have moved.